File: ruby/embed.rb
# Embed an existing UN/CEFACT CII XML into an existing PDF, producing a
# Factur-X compliant PDF/A-3 hybrid invoice, using the InvoiceXML API.
#
# Your PDF keeps its exact layout, fonts, and branding; the API promotes the
# container to PDF/A-3 and attaches the XML as factur-x.xml. The XML is
# validated with the full /v1/validate/facturx rule set first, so a
# non-compliant invoice never leaves the API.
#
# Get API key: https://www.invoicexml.com/account/authentication
# Docs:        https://www.invoicexml.com/docs/api/embed/facturx
#
# Ruby 3.x, standard library only. Set the INVOICEXML_API_KEY
# environment variable to the raw key, without the "Bearer " prefix.
#
# Usage: ruby embed.rb [path/to/invoice.pdf] [path/to/factur-x.xml]

require "net/http"
require "uri"

BASE_URL = "https://api.invoicexml.com"
API_KEY  = ENV.fetch("INVOICEXML_API_KEY")

def api_http(uri)
  http = Net::HTTP.new(uri.host, uri.port)
  http.use_ssl = true
  http.open_timeout = 10
  http.read_timeout = 120
  http
end

pdf_path = ARGV.fetch(0, "invoice.pdf")
xml_path = ARGV.fetch(1, "factur-x.xml")

uri = URI("#{BASE_URL}/v1/embed/facturx")
request = Net::HTTP::Post.new(uri)
request["Authorization"] = "Bearer #{API_KEY}"

# Two file parts in one multipart request: "pdf" is the visual layer,
# "xml" is the CII payload. skipValidation is optional: "true" packages the
# XML without EN 16931 business-rule checks, structural checks still run.
response = File.open(pdf_path, "rb") do |pdf|
  File.open(xml_path, "rb") do |xml|
    request.set_form(
      [["pdf", pdf], ["xml", xml], ["skipValidation", "false"]],
      "multipart/form-data"
    )
    api_http(uri).request(request)
  end
end

unless response.is_a?(Net::HTTPSuccess)
  # 400 with errorCode 4001 lists every failed business rule;
  # 4017 means BT-24 is not an official Factur-X profile URN.
  warn "InvoiceXML API error #{response.code}: #{response.body}"
  exit 1
end

File.binwrite("invoice-facturx.pdf", response.body)
puts "Saved invoice-facturx.pdf (#{response.body.bytesize} bytes)"

File: nodejs/embed.js
// Embed an existing UN/CEFACT CII XML into an existing PDF, producing a
// Factur-X compliant PDF/A-3 hybrid invoice.
//
// Your PDF keeps its exact layout, fonts, and branding; the API promotes the
// container to PDF/A-3 and attaches the XML as factur-x.xml. The XML is
// validated with the full /v1/validate/facturx rule set first, so a
// non-compliant invoice never leaves the API.
//
// Get API key: https://www.invoicexml.com/account/authentication
// Docs:        https://www.invoicexml.com/docs/api/embed/facturx
//
// Requires Node.js 18+. No dependencies.

const fs = require('fs');

// Raw key only, without the "Bearer " prefix (it is added below).
const apiKey = 'YOUR_API_KEY';

(async () => {
    const form = new FormData();
    form.append('pdf', new Blob([fs.readFileSync('invoice.pdf')], { type: 'application/pdf' }), 'invoice.pdf');
    form.append('xml', new Blob([fs.readFileSync('factur-x.xml')], { type: 'application/xml' }), 'factur-x.xml');
    // Optional. 'true' packages the XML without EN 16931 business-rule checks;
    // structural checks (CII root, BT-24 profile URN, XSD) still run.
    form.append('skipValidation', 'false');

    const response = await fetch('https://api.invoicexml.com/v1/embed/facturx', {
        method: 'POST',
        headers: { Authorization: `Bearer ${apiKey}` },
        body: form,
    });

    if (!response.ok) {
        // 400 with errorCode 4001 lists every failed business rule;
        // 4017 means BT-24 is not an official Factur-X profile URN.
        console.error(`InvoiceXML API error ${response.status}: ${await response.text()}`);
        process.exit(1);
    }

    const buffer = Buffer.from(await response.arrayBuffer());
    fs.writeFileSync('invoice-facturx.pdf', buffer);
    console.log(`Saved invoice-facturx.pdf (${buffer.length} bytes)`);
})();
